lib/plugins/place_entity.js

```diff
@@ -0,0 +1,175 @@
+      bot._client.write('block_place', {
+        location: pos,
+        direction: vectorToDirection(faceVector),
+        hand: 0,
+        cursorX: dx,
+        cursorY: dy,
+        cursorZ: dz
+      })
+    } else if (bot.supportFeature('blockPlaceHasInsideBlock')) {
+      bot._client.write('block_place', {
+        location: pos,
+        direction: vectorToDirection(faceVector),
+        hand: 0,
+        cursorX: dx,
+        cursorY: dy,
+        cursorZ: dz,
+        insideBlock: false
+      })
+      if (type === 'spawn_egg') {
+        bot.swingArm(undefined, false)
+      } else {
+        bot.swingArm()
+      }
+    }
+
+    if (type === 'boat') {
+      if (bot.supportFeature('useItemWithOwnPacket')) {
+        bot._client.write('use_item', {
+          hand: 36 - bot.heldItem.slot
+        })
+      } else {
+        bot._client.write('block_place', {
+          location: { x: -1, y: -1, z: -1 },
+          direction: -1,
+          heldItem: Item.toNotch(bot.heldItem),
+          cursorX: 0,
+          cursorY: 0,
+          cursorZ: 0
+        })
+      }
+    }
+
+    const dest = pos.plus(faceVector)
+
+    return waitForEntitySpawn(name, dest)
+  }
+
+  async function placeEntity (referenceBlock, faceVector) {
+    return await placeEntityWithOptions(referenceBlock, faceVector, {})
+  }
+
+  function waitForEntitySpawn (name, placePosition) {
+    const maxDistance = name === 'bat' ? 4 : name === 'boat' ? 3 : 2
+    let mobName = name
+    if (name === 'end_crystal') {
+      if (bot.supportFeature('enderCrystalNameEndsInErNoCaps')) {
+        mobName = 'ender_crystal'
+      } else if (bot.supportFeature('entityNameLowerCaseNoUnderscore')) {
+        mobName = 'endercrystal'
+      } else if (bot.supportFeature('enderCrystalNameNoCapsWithUnderscore')) {
+        mobName = 'end_crystal'
+      } else {
+        mobName = 'EnderCrystal'
+      }
+    } else if (name === 'boat') {
+      mobName = bot.supportFeature('entityNameUpperCaseNoUnderscore') ? 'Boat' : 'boat'
+    } else if (name === 'armor_stand') {
+      if (bot.supportFeature('entityNameUpperCaseNoUnderscore')) {
+        mobName = 'ArmorStand'
+      } else if (bot.supportFeature('entityNameLowerCaseNoUnderscore')) {
+        mobName = 'armorstand'
+      } else {
+        mobName = 'armor_stand'
+      }
+    }
+
+    return new Promise((resolve, reject) => {
+      function listener (entity) {
+        const dist = entity.position.distanceTo(placePosition)
+        if (entity.name === mobName && dist < maxDistance) {
+          resolve(entity)
+        }
+        bot.off('entitySpawn', listener)
+      }
+
+      setTimeout(() => {
+        bot.off('entitySpawn', listener)
+        reject(new Error('placeEntity timed out'))
+      }, 5000) // reject after 5s
+
+      bot.on('entitySpawn', listener)
+    })
+  }
+
+  bot.placeEntity = callbackify(placeEntity)
+  bot._placeEntityWIthOptions = placeEntityWithOptions
+}
+
+function vectorToDirection (v) {
```

Comment:
don't we have this function elsewhere already ?

Reply:
yeah, in placeBlock :p maybe should go in util?

Reply:
yeah something like that

Reply:
Done